 Multiple Linear Regression 

## Implementing Multiple Linear Regression Using Gradient Descent And Normal Equations Method.



In [2]:
# Run this code cell using Shift + Enter before moving further
import numpy as np
import pandas as pd
import matplotlib
from matplotlib import pyplot as plt

# Loading the data set.

In the below code cell, you will load the data using python pandas library as done in the previous programming assignment. Use `pd.read_csv('File url ', header=None,.... )` with the value of `header=None`,`delim_whitespace=True`,`names=names`,`na_values='?'` as attributes. The url for the .data file is https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data. 

In [3]:
# After completing the code in this code cell, run this code cell before moving further.
names =[
    'CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 
    'AGE',  'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'PRICE'
]


# Write the pandas command to read the csv file from the website into df(<pd.Dataframe>)
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data',names=names,delim_whitespace=True)
print(df)


        CRIM    ZN  INDUS  CHAS    NOX  ...    TAX  PTRATIO       B  LSTAT  PRICE
0    0.00632  18.0   2.31     0  0.538  ...  296.0     15.3  396.90   4.98   24.0
1    0.02731   0.0   7.07     0  0.469  ...  242.0     17.8  396.90   9.14   21.6
2    0.02729   0.0   7.07     0  0.469  ...  242.0     17.8  392.83   4.03   34.7
3    0.03237   0.0   2.18     0  0.458  ...  222.0     18.7  394.63   2.94   33.4
4    0.06905   0.0   2.18     0  0.458  ...  222.0     18.7  396.90   5.33   36.2
..       ...   ...    ...   ...    ...  ...    ...      ...     ...    ...    ...
501  0.06263   0.0  11.93     0  0.573  ...  273.0     21.0  391.99   9.67   22.4
502  0.04527   0.0  11.93     0  0.573  ...  273.0     21.0  396.90   9.08   20.6
503  0.06076   0.0  11.93     0  0.573  ...  273.0     21.0  396.90   5.64   23.9
504  0.10959   0.0  11.93     0  0.573  ...  273.0     21.0  393.45   6.48   22.0
505  0.04741   0.0  11.93     0  0.573  ...  273.0     21.0  396.90   7.88   11.9

[506 rows x 14 

Now, instead of taking just 1 feature i.e 'RM' as we did in Simple Linear Regression, we will use 4 features which are 'LSTAT', 'DIS', 'RM' and 'INDUS'. We choose only these features because they are almost on similar scale. So, there is no need for feature scaling. Fetching the values of 5 columns into a smaller dataframe (say) `df1` from `df`. This step is also same as done in the  previous programming assignment.  There we fetched only 2 columns but here we need 5 columns.We need the values in the 'PRICE', 'LSTAT', 'DIS', 'RM' and 'INDUS' columns.

In [4]:
# TODO Q02
# Save values from columns 'PRICE', 'LSTAT', 'DIS', 'RM', 'INDUS' into df1(<pd.Dataframe>) from df 
df1 = df[['PRICE', 'LSTAT', 'DIS', 'RM', 'INDUS']]
# Remove nan values from df1 and save into df2(<pd.Dataframe>)
df2 = df1.dropna()
# Check the shape of df2. It should be (506,5)
print("shapde of df2: ", df2.shape)


shapde of df2:  (506, 5)


In [5]:

# Create a vector y(<np.ndarray>) having the values of 'PRICE' column and matrix X(<np.ndarray>) 
# having the values of 'LSTAT', 'DIS', 'RM' and 'INDUS' columns
X = df2[['LSTAT', 'DIS', 'RM', 'INDUS']].to_numpy()
y = df2["PRICE"].to_numpy()
# Check the shape of X and y vectors.
print("shapde of X: ", X.shape)
print("shapde of y: ", y.shape)

shapde of X:  (506, 4)
shapde of y:  (506,)


Reshape y to be rank 2. After checking the shape of `X` and `y` in the above code cell, we see that `X` is already rank 2 and `y` are rank 1 matrix. Before moving ahead, convert `y` to be rank 2 matrix. For example, I would use the command `y = y.reshape(y.shape[0],1)` to reshape.

In [6]:

# Reshape y into a rank 2 matrix y_2d(<np.ndarray>)
y_2d = y.reshape(y.shape[0],1)
print("shapde of y_2d: ", y_2d.shape)

shapde of y_2d:  (506, 1)


Calculating the value of `N` i.e. number of training examples. 
Hint: Value of `N` is equal to the number of rows in either `X` or `y` matrix which can be accessed using numpy shape command.

In [7]:

# Save number of training examples into N and print it
N = y.shape[0]
print("N: ", N)

N:  506


In [17]:

# Create a column of ones and name it as ones, then append it to left side X using hstack, 
# name the new matrix X_1(<np.ndarray>)
ones = np.ones((N,1))
X_1 = np.hstack((ones,X))
print("X_1:", X_1)

# Shape of X_1 should be (506,5)
print("shapde of X_1: ", X_1.shape)

X_1: [[ 1.      4.98    4.09    6.575   2.31  ]
 [ 1.      9.14    4.9671  6.421   7.07  ]
 [ 1.      4.03    4.9671  7.185   7.07  ]
 ...
 [ 1.      5.64    2.1675  6.976  11.93  ]
 [ 1.      6.48    2.3889  6.794  11.93  ]
 [ 1.      7.88    2.505   6.03   11.93  ]]
shapde of X_1:  (506, 5)


# Cost Function.
Compute the cost: Write the code to compute the cost inside the function. Do not change the function name or function parameters.

In [20]:
def compute_cost(X_1, y_2d, w, N):
    # Write your code in place of ellipsis. Cost can be calculated using a single line of code.
    # Remember w is a vector here.
    

    # Write the cost function
    cost = (((y_2d - np.dot(X_1,w))**2).sum())/(2*N)
    
    return cost

Before moving ahead, ensure that the code you have written to compute the cost is correct. Just run the below cell.

In [47]:
w_testcase=np.zeros((5,1))
cost_verify= compute_cost(X_1, y_2d, w_testcase, N)
print("cost_verify: ", cost_verify)

cost_verify:  296.0734584980237


Your output should be equal to 296.073458498. If it's equal to this, then move ahead. Else, re-check your code and re-verify.

# Gradient Descent
Write the code to perform gradient descent in the function below

In [26]:
def gradient_descent(X_1, y_2d, learning_rate, w, N, num_iters):
    # In place of ellipsis, write the updated value of w0 in temp0 and of w1 in temp1
    # TODO Q08
    # Finish the gradient descent function
    
    for i in range(num_iters):
        # derivative vector is given by : X_train.Transpose *  (( X_train * w)- y ) 
        w = w - learning_rate*(1/N)*np.dot(X_1.T, (np.dot(X_1,w) - y_2d))
       
        
        if(i % 100 == 0):
            # In place of ellipsis, call the cost you just coded above
            cost = compute_cost(X_1,y_2d,w,N)
            # You can uncomment print statements below to see how cost changes, 
            # but please make sure you put prints in comments before submission
            #print("Cost")
            #print(cost)
             
    return w      

Before moving ahead, ensure that your code to update `w` is correct. Run the code cell below.

In [48]:
w_testcase = np.zeros((5,1))
g = gradient_descent(X_1, y_2d, 0.0049, w_testcase, N, 100000)
print("g[0]: ", g[0])

g[0]:  [7.46236374]


The last output should be: 7.46236374. If it matches your out, then your code is likely to be correct. Otherwise, re-check your code and re-verify.

# Integrating the Batch Gradient Descent Function 

Integrating the above function into a single function `multiple_linear_reg_model_gda`: This function uses gradient descent algorithm to minimize the cost.

In [30]:
def multiple_linear_reg_model_gda(X_1, y_2d, learning_rate, N, num_iters):
    #initialize the values of parameter vector w. It should be a column vector of zeros of dimension(d+1,1)
    # TODO 09 
    # Complete the function
    w = np.zeros((X_1.shape[1],1))
    
    # Calculate the initial cost by calling the function you coded above.
    initial_cost = compute_cost(X_1,y_2d,w,N)
    
    # Calculate the optimized value of gradients by calling the gradient_descent function coded above
    w = gradient_descent(X_1,y_2d,learning_rate,w,N,num_iters)
    
    # Calculate the cost with the optimized value of w0 and w1 by calling the cost function.
    final_cost = compute_cost(X_1,y_2d,w,N)

    return w, initial_cost, final_cost

Now, when you have coded `multiple_linear_reg_model_gda` function, you can call this function to find the optimized values of parameters `w`. Before calling the function, set the values of `learning_rate` and `num_iters`. You may have to call this function several number of times with different values of num_iters and learning_rate to find the optimal values of `w`. For some values of `learning_rate`, it may give an error as the values of cost may reach a very high value(infinity) due to overshooting as discussed in the class. 

In [69]:

# Call the multiple_linear_reg_model_gda.
learning_rate = 0.004
num_iters = 500000
results = multiple_linear_reg_model_gda(X_1,y_2d,learning_rate,N,num_iters)
print("w: ", results[0], "\ninitial cost: ", results[1], "\nfinal cost: ", results[2])

# The value of final cost should be 14.3470049896 or nearly this(depending on the values of learning_rate and num_itersations you choose.)

w:  [[ 8.29557681]
 [-0.6646653 ]
 [-0.90385718]
 [ 4.5807493 ]
 [-0.24341693]] 
initial cost:  296.0734584980237 
final cost:  14.345406879911714


# Normal Equation Method
Now, we will be writing the code to find the values of parameters `w` for our multiple linear regression model. This can also be used to cross-check the optimal values of `w` we just found above using method above. These values should be same (or nearly same).
Instead of writing the code for normal equation in one line, you can break this into 3 parts: First calculate `q = inverse of (dot of (X_1.T, X_1))` (these are pseudocode commands, use original numpy commands to calculate `q`). Then `z = dot of (X_1.T, y_2d)` and then `w_vec = dot of (q, z)`. Here, `w_vec` is vector of dimension (5,1) having two values. Example `w0 = w_vec[0][0]`.

Note : Do not append a column of ones in `X_1` because you have already done this. 

In [44]:
# Write the code below
# Write the normal equation method, save the weights to w_vec(<np.ndarray>)

q = np.linalg.inv(np.dot(X_1.T,X_1))
z = np.dot(X_1.T,y_2d)
w_vec = np.dot(q,z)

In [45]:
print("w_vec: ", w_vec)

w_vec:  [[ 8.29647186]
 [-0.66467285]
 [-0.90387731]
 [ 4.58064345]
 [-0.24342177]]


The values of `w` you just got above should be approximately same as the ones you got using `multiple_linear_reg_model_gda`.
